### PR TITLE
Feature/tpa 0013 admin unapproved accounts

### PR DESCRIPTION
**Pull request by DeveloperX104**
Un approved users are being rendered in the `admin/view/users/unapprove` endpoint.

### CODE REVIEW

File: src/pages/AdminPages/UnapprovedAccounts.js
Comment (2kama) [inline src/pages/AdminPages/UnapprovedAccounts.js]: I feel we should follow a pattern where every page is made a folder with an `index.jsx` file in it. Then subsequent component that will only be utilized by that page will be added to the same folder.

Make this `UnapprovedAccounts/index.jsx`

File: src/store/sagas/requests/user.js
Comment (2kama) [inline src/store/sagas/requests/user.js]: Remember I said `newUser` is not a reliable collection. Instead search for users in the `users` collection whole have their `isApproved` field set to false in the `private/info` subcollection. Then use their `email` to search for their document in the `newUser` collection

File: src/pages/AdminPages/UnapprovedAccounts.js
Comment (2kama) [inline src/pages/AdminPages/UnapprovedAccounts.js]: We might have to rework on this. The admin has to be able to able to set some properties for the user before approving. The client doesn't want to just approving all users at once. he wants to be able to read through each registration and decide to approve or deny.

File: src/pages/AdminPages/UnapprovedAccounts.js
Comment (2kama) [inline src/pages/AdminPages/UnapprovedAccounts.js]: This dispatch call is running multiple times... This should only query the database once on page load. To update the page, the page has to be reloaded

File: src/store/sagas/requests/user.js
Comment (2kama) [inline src/store/sagas/requests/user.js]: Admin getting unapproved users is not a user specific operation... a new file should be created for these. Maybe call it `adminQueryUsers.js`

File: src/store/sagas/handlers/user.js
Comment (2kama) [inline src/store/sagas/handlers/user.js]: Admin getting unapproved users is not a user specific operation... a new file should be created for these. Maybe call it `adminQueryUsers.js` [Part II]

File: src/store/sagas/handlers/user.js
Comment (2kama) [inline src/store/sagas/handlers/user.js]: Set this to push an alert if there's an error.

File: src/pages/AdminPages/UnapprovedAccounts.js
Comment (2kama) [reply src/pages/AdminPages/UnapprovedAccounts.js]: Screen Shot 2021-07-22 at 11 17 11 PM

Comment (2kama) [reply src/pages/AdminPages/UnapprovedAccounts.js]: And this is a very bad technique for firebase... Remember you get billed on document reads?
Screen Shot 2021-07-22 at 11 21 24 PM

Comment (2kama) [reply src/pages/AdminPages/UnapprovedAccounts.js]: The spike there, is when you were running this code... 

If you were trying to give a realtime view of the collection, a `snapshot` is a better way to do it. That way we only get a document read when a document is added or removed from the collection.

File: src/pages/AdminPages/UnapprovedAccounts.js
Comment (2kama) [inline src/pages/AdminPages/UnapprovedAccounts.js]: 1. You are presuming that the admin will not first click / accidentally choose something else before choosing the right role...You need to set the other roles to false when he picks  any role.

2. superadmin is also an admin so both will be true when superadmin is chosen. an affiliate is also a user so both will be true when affiliate is chosen

File: src/pages/AdminPages/UnapprovedAccounts.js
Comment (DeveloperX104) [reply src/pages/AdminPages/UnapprovedAccounts.js]: Alright, thanks!

File: src/store/sagas/requests/user.js
Comment (DeveloperX104) [reply src/store/sagas/requests/user.js]: ooh okay.

File: src/pages/AdminPages/UnapprovedAccounts.js
Comment (DeveloperX104) [reply src/pages/AdminPages/UnapprovedAccounts.js]: Yh, so here the user can click on the approve button which opens a modal, the admin can now set properties inside that modal. I'd need you to give me the data that the admin will be setting here also.

File: src/pages/AdminPages/UnapprovedAccounts.js
Comment (DeveloperX104) [reply src/pages/AdminPages/UnapprovedAccounts.js]: Thanks, will fix this.

File: src/store/sagas/requests/user.js
Comment (DeveloperX104) [reply src/store/sagas/requests/user.js]: ok

File: src/store/sagas/handlers/user.js
Comment (DeveloperX104) [reply src/store/sagas/handlers/user.js]: Alright, thanks.

File: src/store/sagas/handlers/user.js
Comment (DeveloperX104) [reply src/store/sagas/handlers/user.js]: Ok

File: src/pages/AdminPages/UnapprovedAccounts.js
Comment (DeveloperX104) [reply src/pages/AdminPages/UnapprovedAccounts.js]: Alright, thanks for this.

File: src/pages/AdminPages/UnapprovedAccounts.js
Comment (2kama) [reply src/pages/AdminPages/UnapprovedAccounts.js]: `role` and `ROI` and `assignedTrader` ...that's all i can think of now.

`assignedTrader` is supposed to be a select field that shows the list of all the traders registered on the app. Then the admin can pick one of the Traders to assign to the user.

NOTE: if the account is made an admin or a trader; `ROI` and `assignedTrader` will not be needed....these fields will be hidden bcos an admin and Trader cannot have an investment portfolio, so ROI is useless and since they don't have an investment portfolio, they can't be assigned to a Trader

Then once, set `isApproved` to `true`, fix the data from `newUser` collection into their respective field in the `user` collection and delete that document from the `newUser` collection.

forget about emailVerification, once the user is approved, the hash page changes to the user clicking a button to send an email verification link (this is already done)

File: src/store/sagas/requests/user.js
Comment (DeveloperX104) [reply src/store/sagas/requests/user.js]: So after looking at it again, I feel the we shouldn't create anything in the `users` collection until the account is approved, it makes it easier and better to work with. So when the user signs up, they have their data in the `newUser` collection then when the account is approved the data is transferred to the `users` collection

File: src/pages/AdminPages/UnapprovedAccounts.js
Comment (DeveloperX104) [reply src/pages/AdminPages/UnapprovedAccounts.js]: Alright thanks.

File: src/pages/AdminPages/UnapprovedAccounts/index.js
Comment (2kama) [inline src/pages/AdminPages/UnapprovedAccounts/index.js]: The default value will be `user`

File: src/store/sagas/requests/adminQueryUsers.js
Comment (2kama) [inline src/store/sagas/requests/adminQueryUsers.js]: You have to get `users` that have `isApproved` set to `false`, not all the `users`. There is a `where` method in firebase which you can use to filter document gets. Imagine having only 1 unapproved user and you'd have to get over 500 documents just for that 1 user. Pay Per document read

File: src/store/sagas/requests/adminQueryUsers.js
Comment (2kama) [inline src/store/sagas/requests/adminQueryUsers.js]: This is not a blocker, but why do you need to reverse?

File: src/pages/AdminPages/UnapprovedAccounts/index.js
Comment (2kama) [inline src/pages/AdminPages/UnapprovedAccounts/index.js]: Let's use `Formik` like I did in Login and Register to process all our forms... Thanks

File: src/store/sagas/requests/adminQueryUsers.js
Comment (DeveloperX104) [reply src/store/sagas/requests/adminQueryUsers.js]: I feel having the recent data at the top is better.

File: src/pages/AdminPages/UnapprovedAccounts/index.js
Comment (DeveloperX104) [reply src/pages/AdminPages/UnapprovedAccounts/index.js]: Alright

File: src/store/sagas/requests/adminQueryUsers.js
Comment (2kama) [reply src/store/sagas/requests/adminQueryUsers.js]: Well, there really is no need for it. The data you're getting is arranged in alphabetical order of the `uid` ... and u know `uid` is a mayhem of alpha-numeric irrespective of who came first.

If you were to sort the data somehow (maybe by `createdAt` field), the oldest entry should be first...not the latest. But even this will need you to start creating composite indexes on firebase...so...on the long run, not needed.

but again, this is no blocker at all.

Comment (DeveloperX104) [reply src/store/sagas/requests/adminQueryUsers.js]: Thanks for the heads up!

File: src/store/sagas/requests/user.js
Comment (2kama) [reply src/store/sagas/requests/user.js]: Again, that can't work, remember `newUser` gets populated even if the register fails.... that means there could be data in newUser that doesn't belong to any registered user. It's an unreliable collection.

File: src/store/sagas/requests/adminQueryUsers.js
Comment (DeveloperX104) [reply src/store/sagas/requests/adminQueryUsers.js]: The `user` collection itself doesn't have a `isApproved` field.

Comment (2kama) [reply src/store/sagas/requests/adminQueryUsers.js]: You wanna check that again? check the `cloud-functions` folder in the `functions` folder. Or check the firestore database

File: src/store/sagas/requests/adminQueryUsers.js
Comment (2kama) [inline src/store/sagas/requests/adminQueryUsers.js]: You need to fix this... use a `where` method in firebase so you're only reading documents that you need and not all the documents

File: src/store/sagas/requests/adminQueryUsers.js
Comment (2kama) [reply src/store/sagas/requests/adminQueryUsers.js]: ???

Comment (DeveloperX104) [reply src/store/sagas/requests/adminQueryUsers.js]: Ooh thanks, just understamding what you meant here

Comment (DeveloperX104) [reply src/store/sagas/requests/adminQueryUsers.js]: I'm having issue trying to chain the query

Comment (2kama) [reply src/store/sagas/requests/adminQueryUsers.js]: What's the issue?

Comment (2kama) [reply src/store/sagas/requests/adminQueryUsers.js]: https://firebase.google.com/docs/firestore/query-data/queries#collection-group-query

Comment (DeveloperX104) [reply src/store/sagas/requests/adminQueryUsers.js]: Don't you think the `isApproved`  field should be inside the `users` collection directly and not in a subcollection.

Comment (2kama) [reply src/store/sagas/requests/adminQueryUsers.js]: all the data that is inside the `private/info` document are data that the user cannot edit. the data that is directly in the `users` document is the data the user can edit.

We wouldn't want the user to be able to edit his `wallet`, `approval` status, `uid`, `role` etc...

File: src/pages/AdminPages/UnapprovedAccounts/index.js
Comment (2kama) [inline src/pages/AdminPages/UnapprovedAccounts/index.js]: This component is doing a lot. Let's trim this down.

Make each `unapproved user` row a component...that way the logic won't be so complex. And try utilizing the `Form Components` built. I've used it a couple of times so you can get an idea of how it works. It'll make Form functions neater and more readable

**Review by 2kama — DISMISSED** [review]

File: src/pages/AdminPages/UnapprovedAccounts/index.js
Comment (DeveloperX104) [reply src/pages/AdminPages/UnapprovedAccounts/index.js]: Thanks, I'd do this now

File: src/components/Modal/BaseModal.js
Comment (2kama) [inline src/components/Modal/BaseModal.js]: let this be a prop.... we might need a `md` or `sm` modal

File: src/pages/AdminPages/UnapprovedAccounts/SelectedUserModal.js
Comment (2kama) [inline src/pages/AdminPages/UnapprovedAccounts/SelectedUserModal.js]: I still have issues with the complexity of this file

File: src/components/Modal/BaseModal.js
Comment (2kama) [inline src/components/Modal/BaseModal.js]: What was your reason for updating the props of your modal?

Comment (2kama) [reply src/components/Modal/BaseModal.js]: Not that I don't like this...it looks cleaner

File: src/components/Modal/BaseModal.js
Comment (2kama) [reply src/components/Modal/BaseModal.js]: add a `size` props

File: src/components/Modal/BaseModal.js
Comment (DeveloperX104) [reply src/components/Modal/BaseModal.js]: I removed the `doneText` and the `onDone` props as I'm using form here so I do not need those props

File: src/pages/AdminPages/UnapprovedAccounts/SelectedUserModal.js
Comment (DeveloperX104) [reply src/pages/AdminPages/UnapprovedAccounts/SelectedUserModal.js]: Most of what is in this file are functions, I've tried making the file smaller

File: src/pages/AdminPages/UnapprovedAccounts/UnapprovedUsers.js
Comment (2kama) [inline src/pages/AdminPages/UnapprovedAccounts/UnapprovedUsers.js]: Why are we back to using `UnapprovedUsers` ?

File: src/store/sagas/requests/adminQueryUsers.js
Comment (2kama) [inline src/store/sagas/requests/adminQueryUsers.js]: This function shouldn't be an async function. We really don't need to worry if the `firstName`, `lastName` and `phone` is updated bcos the user can still do this himself if something wrong happens on the admin end.

But we are particular about the `private` collection being updated bcos that is the whole purpose of the update user process.... so we need the promise returned from updating the private collection --- we need to know if it failed or pass.

with this technique, we'll never know bcos you're not even returning the promise to the `handler`.

Comment (2kama) [reply src/store/sagas/requests/adminQueryUsers.js]: The implementation from my last commit on this function is better aligned to this.

Comment (2kama) [reply src/store/sagas/requests/adminQueryUsers.js]:
```
export const alterUser = (data) => {

    db.doc(`users/${data.uid}`).update({
        firstName: data.firstName,
        lastName: data.lastName,
        phone: data.phone
    })
    
    return db.doc(`users/${data.uid}/private/info`).update({
        role: data.role,
        isApproved: true,
        ROI: data.roi < 0 ? 0 : data.roi,
        assignedTrader: data.role.isUser ? data.assignedTrader : "",
        affiliateCode: data.role.isAffliate ? data.affiliateCode : "",
        affiliate: data.affiliate
    })
}
```

Change the whole func to this...

File: functions/cloud_functions/user.js
Comment (2kama) [inline functions/cloud_functions/user.js]: Delete this line... we are not doing anything to `newUser` but populating and reading it for approvals

File: functions/index.js
Comment (2kama) [inline functions/index.js]: Not a needed cloud function.... `userApproved`

File: src/pages/AdminPages/UnapprovedAccounts/SelectedUserModal.js
Comment (2kama) [inline src/pages/AdminPages/UnapprovedAccounts/SelectedUserModal.js:29]: Take these off... Not necessary.... They aren't a `mustFill` field.

File: functions/cloud_functions/user.js
Comment (2kama) [inline functions/cloud_functions/user.js]: we don't need this cloud function...pay no attention to `newUser` collection, it's just a dumping group.

Besides, anytime an admin updates a user, even though he wasn't approving a user, this function will run...so it's a waste of read and write activities.

File: src/pages/AdminPages/UnapprovedAccounts/SelectedUserModal.js
Comment (2kama) [inline src/pages/AdminPages/UnapprovedAccounts/SelectedUserModal.js]: What is this for?

File: src/pages/AdminPages/UnapprovedAccounts/SelectedUserModal.js
Comment (2kama) [inline src/pages/AdminPages/UnapprovedAccounts/SelectedUserModal.js]: Why are we importing this twice?

File: src/pages/AdminPages/UnapprovedAccounts/SelectedUserModal.js
Comment (DeveloperX104) [reply src/pages/AdminPages/UnapprovedAccounts/SelectedUserModal.js]: This is the dispatch function that actually calls the generator that approves the user account.

File: src/store/sagas/requests/adminQueryUsers.js
Comment (2kama) [inline src/store/sagas/requests/adminQueryUsers.js]: Change to this...

```
db.doc(`users/${data.uid}`).update({
        firstName: data.firstName,
        lastName: data.lastName,
        phone: data.phone
    })

return db.doc(`users/${data.uid}/private/info`).update(updatedData)
```

Comment (2kama) [reply src/store/sagas/requests/adminQueryUsers.js]: We want the promise from updating the `private/info` not the `users` document

File: src/pages/AdminPages/UnapprovedAccounts/SelectedUserModal.js
Comment (2kama) [inline src/pages/AdminPages/UnapprovedAccounts/SelectedUserModal.js]: Return this function back to how it was previously

File: src/store/sagas/requests/adminQueryUsers.js
Comment (2kama) [inline src/store/sagas/requests/adminQueryUsers.js]: Change all these to : 


```
export const alterUser = (data) => {

    db.doc(`users/${data.uid}`).update({
        firstName: data.firstName,
        lastName: data.lastName,
        phone: data.phone
    })
    
    return db.doc(`users/${data.uid}/private/info`).update({
        role: data.role,
        isApproved: true,
        ROI: data.roi < 0 ? 0 : data.roi,
        assignedTrader: data.role.isUser ? data.assignedTrader : "",
        affiliateCode: data.role.isAffliate ? data.affiliateCode : "",
        referralCode: data.referralCode
    })
}
```

File: functions/cloud_functions/user.js
Comment (2kama) [inline functions/cloud_functions/user.js]: change to `referralCode`

File: src/store/sagas/requests/adminQueryUsers.js
Comment (2kama) [reply src/store/sagas/requests/adminQueryUsers.js]: Over here

**Review by 2kama — APPROVED** [review]

**Pull request merged**
